### PR TITLE
NFRR Audit Collector and API Implementation

### PR DESCRIPTION
**Pull request by nameisaravind**
This PR comprises of, NFRR Audit Collector and its API Implementation. This collector collects audit statuses for all the available dashboards and store the audit data in audit_results collection. APIs through which can get audit results.

### CODE REVIEW

**Comment by nameisaravind** [comment]
NFRR API Endpoints, Response time and response data size is reviewed with Topo, asked to add pagination support on few.

**Comment by kumarvarun1252** [comment]
@nameisaravind please update to the latest code

File: collectors/audit/nfrr/src/test/java/com/capitalone/dashboard/collector/AuditCollectorTaskTest.java [hunk 1]
```diff
@@ -0,0 +1,86 @@
...
+                assertTrue(auditResult.getDashboardId().equals(dashboard.getId()));
+                assertTrue(auditResult.getDashboardTitle().equalsIgnoreCase(dashboard.getTitle()));
+                assertTrue(auditResult.getTimestamp() == END_DATE);
+                auditResults.add(auditResult);
+                assertNotNull(auditResults);
+            } catch (Exception e) {
+                e.getLocalizedMessage();
+            }
+            if (!auditResults.isEmpty()) {
+                mockstatusRepository.save(auditResults);
```
Comment (tabladrum) [inline collectors/audit/nfrr/src/test/java/com/capitalone/dashboard/collector/AuditCollectorTaskTest.java]: Isn't the save something that you should test?

File: collectors/audit/nfrr/src/main/java/com/capitalone/dashboard/collector/AuditCollector.java [hunk 2]
```diff
@@ -0,0 +1,40 @@
...
+        options.put(INSTANCE_URL, "");
+        options.put(JOB_URL, "");
+        options.put(JOB_NAME, "");
+        protoType.setAllFields(options);
+        protoType.setUniqueFields(options);
+        return protoType;
+    }
+
+    public List<String> getBuildServers() {
+        return buildServers;
```
[the remark this answers is not in the document]
Comment (nameisaravind) [reply collectors/audit/nfrr/src/main/java/com/capitalone/dashboard/collector/AuditCollector.java]: these properties are not needed and removed

File: collectors/audit/nfrr/src/main/java/com/capitalone/dashboard/collector/AuditCollectorTask.java [hunk 3]
```diff
@@ -0,0 +1,128 @@
...
+    }
+
+    public AuditCollectorTask (TaskScheduler taskScheduler) {
+        super(taskScheduler, "AuditCollector");
+    }
+
+    @Override
+    public void collect(AuditCollector collector) {
+        LOGGER.info("NFRR Audit Collector pulls all the dashboards");
+        Iterable<Dashboard> dashboards = dashboardRepository.findAll(new Sort(Sort.Direction.ASC, "title"));
```
[the remark this answers is not in the document]
Comment (nameisaravind) [reply collectors/audit/nfrr/src/main/java/com/capitalone/dashboard/collector/AuditCollectorTask.java]: Changes done to retrieve only team dashboards.

File: collectors/audit/nfrr/src/main/java/com/capitalone/dashboard/collector/AuditCollectorTask.java [hunk 4]
```diff
@@ -0,0 +1,128 @@
...
+
+        List<AuditResult> auditResults = getAuditResults(dashboards);
+        if (!auditResults.isEmpty()) {
+            try {
+                AuditCollectorUtil.clearAuditResultRepo(auditResultRepository);
```
[the remark this answers is not in the document]
Comment (nameisaravind) [reply collectors/audit/nfrr/src/main/java/com/capitalone/dashboard/collector/AuditCollectorTask.java]: Updated as commented.

File: collectors/audit/nfrr/src/main/java/com/capitalone/dashboard/collector/AuditCollectorTask.java [hunk 5]
```diff
@@ -0,0 +1,128 @@
...
+                auditResultRepository.save(auditResults);
+                AuditCollectorUtil.clearAuditResults();
+                LOGGER.info("NFRR Audit Collector executed successfully");
+            } catch (Exception e) {
+                LOGGER.error("Error while saving audit status data to database", e.getMessage());
```
[the remark this answers is not in the document]
Comment (nameisaravind) [reply collectors/audit/nfrr/src/main/java/com/capitalone/dashboard/collector/AuditCollectorTask.java]: Updated as commented.

File: collectors/audit/nfrr/src/main/java/com/capitalone/dashboard/collector/AuditCollectorTask.java [hunk 6]
```diff
@@ -0,0 +1,128 @@
...
+                        auditBeginDateTimeStamp, auditEndDateTimeStamp);
+
+                LOGGER.info("NFRR Audit Collector adding audit results for the dashboard : " + dashboard.getTitle());
+                AuditCollectorUtil.addAuditResultByAuditType(dashboard, auditMap, cmdbRepository, auditEndDateTimeStamp);
+            }
+            catch(HttpClientErrorException hce){
+                LOGGER.error("Http Error while calling audit api service for the dashboard - " + dashboard.getTitle());
+            }
+            catch (Exception e) {
+                LOGGER.error("Error while calling audit api service for the dashboard - " + dashboard.getTitle());
```
[the remark this answers is not in the document]
Comment (nameisaravind) [reply collectors/audit/nfrr/src/main/java/com/capitalone/dashboard/collector/AuditCollectorTask.java]: Updated as commented.

File: collectors/audit/nfrr/src/main/java/com/capitalone/dashboard/collector/AuditCollectorUtil.java [hunk 7]
```diff
@@ -0,0 +1,490 @@
...
+     * Do basic audit check - configuration, collector error, no data
+     */
+    private static Audit doBasicAuditCheck(JSONArray jsonArray, JSONArray global, AuditType auditType) {
+        Audit audit = new Audit();
+        audit.setType(auditType);
+        if (!isConfigured(auditType, global)) {
+            audit.setDataStatus(DataStatus.NOT_CONFIGURED);
+            audit.setAuditStatus(AuditStatus.NA);
+            return audit;
+        } else if (jsonArray == null || CollectionUtils.isEmpty(jsonArray)) {
```
[the remark this answers is not in the document]
Comment (nameisaravind) [reply collectors/audit/nfrr/src/main/java/com/capitalone/dashboard/collector/AuditCollectorUtil.java]: Updated as commented.

File: collectors/audit/nfrr/src/main/java/com/capitalone/dashboard/collector/AuditCollectorUtil.java [hunk 8]
```diff
@@ -0,0 +1,490 @@
...
+            audit.setAuditStatus(AuditStatus.NA);
+            audit.setDataStatus(DataStatus.NO_DATA);
+            return audit;
+        } else if (isCollectorError(jsonArray)) {
+            audit.setAuditStatus(AuditStatus.NA);
+            audit.setDataStatus(DataStatus.ERROR);
+            return audit;
+        } else {
```
[the remark this answers is not in the document]
Comment (nameisaravind) [reply collectors/audit/nfrr/src/main/java/com/capitalone/dashboard/collector/AuditCollectorUtil.java]: else removed

File: collectors/audit/nfrr/src/main/java/com/capitalone/dashboard/collector/AuditCollectorUtil.java [hunk 9]
```diff
@@ -0,0 +1,490 @@
...
+    }
+
+    /**
+     * Check for dashboard audit type configuration
+     */
+    private static boolean isConfigured(AuditType auditType, JSONArray jsonArray) {
+        if (auditType.equals(AuditType.CODE_REVIEW)) {
+            return (jsonArray.toJSONString().contains("DASHBOARD_REPO_CONFIGURED") ? true : false);
+        } else if (auditType.equals(AuditType.CODE_QUALITY)) {
+            return (jsonArray.toJSONString().contains("DASHBOARD_CODEQUALITY_CONFIGURED") ? true : false);
```
[the remark this answers is not in the document]
Comment (nameisaravind) [reply collectors/audit/nfrr/src/main/java/com/capitalone/dashboard/collector/AuditCollectorUtil.java]: else's removed

File: collectors/audit/nfrr/src/main/java/com/capitalone/dashboard/collector/AuditCollectorUtil.java [hunk 10]
```diff
@@ -0,0 +1,490 @@
...
+                auditJO.stream().map(aj -> audit.getAuditStatusCodes().add((String) aj));
+                boolean ok = false;
+                for (Object s : auditJO) {
+                    String status = (String) s;
+                    audit.getAuditStatusCodes().add(status);
+                    if ("CODE_QUALITY_AUDIT_OK".equalsIgnoreCase(status)) {
+                        ok = true;
+                        break;
+                    }
+                    if ("CODE_QUALITY_DETAIL_MISSING".equalsIgnoreCase(status)) {
```
[the remark this answers is not in the document]
Comment (nameisaravind) [reply collectors/audit/nfrr/src/main/java/com/capitalone/dashboard/collector/AuditCollectorUtil.java]: Audit Status class constants used

File: collectors/audit/nfrr/src/main/java/com/capitalone/dashboard/collector/AuditCollectorUtil.java [hunk 11]
```diff
@@ -0,0 +1,490 @@
...
+     */
+    private static Audit getSecurityAudit(JSONArray jsonArray, JSONArray global) {
+
+        LOGGER.info("NFRR Audit Collector auditing STATIC_SECURITY_ANALYSIS");
+        Audit audit = new Audit();
+        audit.setType(AuditType.STATIC_SECURITY_ANALYSIS);
+        Audit basicAudit;
+        if ((basicAudit = doBasicAuditCheck(jsonArray, global, AuditType.STATIC_SECURITY_ANALYSIS)) != null) {
+            return basicAudit;
+        } else {
```
[the remark this answers is not in the document]
Comment (nameisaravind) [reply collectors/audit/nfrr/src/main/java/com/capitalone/dashboard/collector/AuditCollectorUtil.java]: else's removed

File: collectors/audit/nfrr/src/main/java/com/capitalone/dashboard/collector/AuditCollectorUtil.java [hunk 12]
```diff
@@ -0,0 +1,490 @@
...
+            throw new MBeanServerNotFoundException("No Server Found to run NoFearRelease audit collector");
+        }
+        return settings.getServers().get(0)
+                + HYGIEIA_AUDIT_URL.concat("auditType=ALL"
+                .concat("&title=").concat(dashboard.getTitle() == null ? "" : dashboard.getTitle())
+                .concat("&businessService=").concat(dashboard.getConfigurationItemBusServName() == null ? "" : dashboard.getConfigurationItemBusServName())
+                .concat("&businessApplication=").concat(dashboard.getConfigurationItemBusAppName() == null ? "" : dashboard.getConfigurationItemBusAppName())
+                .concat("&beginDate=").concat(String.valueOf(beginDate))
+                .concat("&endDate=").concat(String.valueOf(endDate)));
+    }
```
[the remark this answers is not in the document]
Comment (nameisaravind) [reply collectors/audit/nfrr/src/main/java/com/capitalone/dashboard/collector/AuditCollectorUtil.java]: Updated with URIEncoder

File: collectors/audit/nfrr/src/main/java/com/capitalone/dashboard/collector/AuditCollectorUtil.java [hunk 13]
```diff
@@ -0,0 +1,490 @@
...
+        ResponseEntity<String> response = restTemplate.exchange(url, HttpMethod.GET, getHeaders(settings), String.class);
+        JSONParser jsonParser = new JSONParser();
+        return (JSONObject) jsonParser.parse(response.getBody());
+    }
+
+    /**
+     * Construct audit api url
+     */
+    @SuppressWarnings("PMD.NPathComplexity")
+    protected static String getAuditAPIUrl(Dashboard dashboard, AuditSettings settings, long beginDate, long endDate) {
```
[the remark this answers is not in the document]
Comment (nameisaravind) [reply collectors/audit/nfrr/src/main/java/com/capitalone/dashboard/collector/AuditCollectorUtil.java]: PMD suppress not needed now, removed.

File: collectors/audit/nfrr/src/test/java/com/capitalone/dashboard/collector/AuditCollectorTaskTest.java (hunk 1 above)
Comment (nameisaravind) [reply collectors/audit/nfrr/src/test/java/com/capitalone/dashboard/collector/AuditCollectorTaskTest.java]: tests added

**Comment by kumarvarun1252** [comment]
Did Squash & Merge

**Pull request merged**